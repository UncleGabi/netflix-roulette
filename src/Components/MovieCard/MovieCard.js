import React, { useEffect, useState } from "react";
import MoreVertIcon from "@mui/icons-material/MoreVert";

import "./MovieCard.scss";

const MovieCard = ({
  id,
  title,
  src,
  genre,
  rating,
  releaseYear,
  duration,
  overview,
  openModal,
  setOpenModal,
  setSelectedMovieData,
}) => {
  const [editDropdownOpen, setEditDropdownOpen] = useState(false);

  useEffect(() => {
    setEditDropdownOpen(false);
  }, [openModal]);

  return (
    <div key={id} className="movie-card">
      <div className="edit-container">
        {editDropdownOpen ? (
          <div className="edit-dropdown">
            <div
              className="close-btn"
              onClick={() => setEditDropdownOpen(false)}
            >
              +
            </div>
            <div
              onClick={() => setOpenModal("Edit")}
              className="edit-dropdown__data"
            >
              Edit
            </div>
            <div
              onClick={() => setOpenModal("Delete")}
              className="edit-dropdown__data"
            >
              Delete
            </div>
          </div>
        ) : (
          <MoreVertIcon
            className="edit-icon"
            onClick={() => setEditDropdownOpen(!editDropdownOpen)}
          />
        )}
      </div>
      <a href="#header">
        <img
          onClick={() =>
            setSelectedMovieData({
              title,
              src,
              genre,
              rating,
              releaseYear,
              duration,
              overview,
            })
          }
          src={`${src}`}
          alt={`${title}`}
        />
      </a>
      <div className="movie-details">
        <div className="movie-title">{title}</div>
        <div className="release-year">{releaseYear}</div>
      </div>
      <div className="genres">{genre.join(" & ")}</div>
    </div>
  );
};

export default MovieCard;
