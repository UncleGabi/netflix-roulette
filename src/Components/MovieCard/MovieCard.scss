.movie-card {
  position: relative;
  margin-top: 0px;

  &:hover {
    .edit-container {
      .edit-icon {
        opacity: 1;
      }
    }
  }

  .edit-container {
    .edit-icon {
      position: absolute;
      top: 20px;
      right: 20px;
      display: flex;
      align-items: center;
      justify-content: center;
      padding: 3px;
      background-color: #232323;
      border-radius: 50%;
      opacity: 0;
      transition: opacity 250ms;
      cursor: pointer;
      width: 30px;
      height: 30px;
    }

    .edit-dropdown {
      position: absolute;
      top: 20px;
      right: 20px;
      width: 50%;
      height: fit-content;
      padding: 30px 0 5px;
      background-color: #232323;

      &__data {
        margin: 5px 0;
        padding: 10px 15px;
        font-size: 20px;
        cursor: pointer;

        &:hover {
          background-color: #f65261;
        }
      }

      & .close-btn {
        position: absolute;
        top: 0px;
        right: 7px;
        font-size: 30px;
        transform: matrix(-0.71, -0.71, -0.71, 0.71, 0, 0);
        cursor: pointer;
      }
    }
  }

  img {
    width: 100%;
    height: 500px;
  }

  .movie-details {
    display: flex;
    justify-content: space-between;
    margin: 15px 0 5px;
    font-size: 14px;

    .movie-title {
      font-weight: 500;
      font-size: 18px;
    }

    .release-year {
      padding: 5px 10px;
      border: 1px solid #979797;
      border-radius: 3px;
    }
  }

  .genres {
    color: #979797;
  }
}
