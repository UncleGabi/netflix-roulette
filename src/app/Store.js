import { configureStore } from "@reduxjs/toolkit";
import MoviesReducer from "../features/MoviesSlice";

export const store = configureStore({
  reducer: {
    movies: MoviesReducer,
  },
});
