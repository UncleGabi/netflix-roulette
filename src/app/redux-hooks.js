import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";

export const useAppDispatch = useDispatch;
export const useAppSelector = useSelector;
export const useAppNavigate = useNavigate;
