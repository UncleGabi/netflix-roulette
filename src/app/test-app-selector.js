const state = {
  movies: [],
  filteredMovies: [],
  selectedMovie: {},
  editedMovie: {},
  searchMovieTitle: "",
};

export const testUseAppSelector = (f) => f(state);
