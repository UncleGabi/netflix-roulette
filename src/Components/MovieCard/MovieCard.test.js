/* eslint-disable testing-library/no-debugging-utils */
/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
import { render, fireEvent, screen } from "@testing-library/react";
import {
  useAppDispatch,
  useAppNavigate,
  useAppSelector,
} from "../../app/redux-hooks";
import renderer from "react-test-renderer";

import MovieCard from "./MovieCard";
import { testUseAppSelector } from "../../app/test-app-selector";
import userEvent from "@testing-library/user-event";
import { BrowserRouter as Router } from "react-router-dom";

const defaultProps = {
  key: "1",
  id: "1",
  title: "Movie title",
  src: "https://movie.com",
  genre: ["Comedy", "Adventure"],
  rating: "8.9",
  releaseYear: new Date("2022-04-04").getFullYear(),
  duration: 120,
  overview: "Description",
  openModal: false,
  setOpenModal: jest.fn(),
  setMovieId: jest.fn(),
};

jest.mock("../../app/redux-hooks");

describe("MovieCard component", () => {
  const dispatch = jest.fn;
  const navigate = jest.fn;

  beforeEach(() => {
    useAppSelector.mockImplementation(testUseAppSelector);
    useAppDispatch.mockImplementation(dispatch);
    useAppNavigate.mockImplementation(navigate);
  });

  afterEach(() => jest.clearAllMocks());
  it("should render correctly", () => {
    const tree = renderer
      .create(
        <Router>
          <MovieCard {...defaultProps} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render edit/delete dropdown and run dispatch when edit is clicked", () => {
    const { container } = render(
      <Router>
        <MovieCard {...defaultProps} />
      </Router>
    );
    const dropdownIcon = container.querySelector(".edit-icon");

    userEvent.click(dropdownIcon);

    expect(container.querySelector(".edit-dropdown")).toBeInTheDocument();

    const editDropdown = container.querySelectorAll(".edit-dropdown__data")[0];

    userEvent.click(editDropdown);

    expect(useAppDispatch).toHaveBeenCalled();
  });

  it("should render edit/delete dropdown and run dispatch when delete is clicked", () => {
    const { container } = render(
      <Router>
        <MovieCard {...defaultProps} />
      </Router>
    );
    const dropdownIcon = container.querySelector(".edit-icon");

    userEvent.click(dropdownIcon);

    expect(container.querySelector(".edit-dropdown")).toBeInTheDocument();

    const editDropdown = container.querySelectorAll(".edit-dropdown__data")[1];

    userEvent.click(editDropdown);

    expect(useAppDispatch).toHaveBeenCalled();
  });

  it("should close edit dropdown when close btn is clicked", () => {
    const { container } = render(
      <Router>
        <MovieCard {...defaultProps} />
      </Router>
    );
    const dropdownIcon = container.querySelector(".edit-icon");

    userEvent.click(dropdownIcon);

    expect(container.querySelector(".edit-dropdown")).toBeInTheDocument();

    const closeBtn = document.querySelector(".close-btn");

    userEvent.click(closeBtn);

    expect(container.querySelector(".edit-dropdown")).not.toBeInTheDocument();
  });

  it("should call dispatch when move img is clicked", () => {
    const { container } = render(
      <Router>
        <MovieCard {...defaultProps} />
      </Router>
    );
    const movieImg = container.querySelector("img");

    userEvent.click(movieImg);

    expect(useAppNavigate).toHaveBeenCalled();
  });
});
