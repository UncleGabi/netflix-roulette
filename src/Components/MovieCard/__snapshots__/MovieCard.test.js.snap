// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard component should render correctly 1`] = `
<div
  className="movie-card"
  onClick={[Function]}
>
  <div
    className="edit-container"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium edit-icon css-i4bv87-MuiSvgIcon-root"
      data-testid="MoreVertIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </div>
  <img
    alt="Movie title"
    onClick={[Function]}
    src="https://movie.com"
  />
  <div
    className="movie-details"
  >
    <div
      className="movie-title"
    >
      Movie title
    </div>
    <div
      className="release-year"
    >
      2022
    </div>
  </div>
  <div
    className="genres"
  >
    Comedy & Adventure
  </div>
</div>
`;
